packages/kbn-openapi-bundler/src/bundler/merge_documents/merge_shared_components.ts

```diff
@@ -6,49 +6,89 @@
  * Side Public License, v 1.
  */
 
-import { OpenAPIV3 } from 'openapi-types';
-import deepEqual from 'fast-deep-equal';
 import chalk from 'chalk';
-import { insertRefByPointer } from '../../utils/insert_by_json_pointer';
-import { ResolvedRef } from '../ref_resolver/resolved_ref';
-import { BundledDocument } from '../bundle_document';
+import deepEqual from 'fast-deep-equal';
+import { OpenAPIV3 } from 'openapi-types';
+import { ResolvedDocument } from '../ref_resolver/resolved_document';
+import { extractByJsonPointer } from '../../utils/extract_by_json_pointer';
+
+const MERGEABLE_COMPONENT_TYPES = [
+  'schemas',
+  'responses',
+  'parameters',
+  'examples',
+  'requestBodies',
+  'headers',
+  'securitySchemes',
+  'links',
+  'callbacks',
+] as const;
 
 export function mergeSharedComponents(
-  bundledDocuments: BundledDocument[]
+  bundledDocuments: ResolvedDocument[]
 ): OpenAPIV3.ComponentsObject {
-  const componentsMap = new Map<string, ResolvedRef>();
   const mergedComponents: Record<string, unknown> = {};
 
-  for (const bundledDocument of bundledDocuments) {
-    mergeRefsToMap(bundledDocument.bundledRefs, componentsMap);
-  }
+  for (const componentsType of MERGEABLE_COMPONENT_TYPES) {
+    const mergedTypedComponents = mergeObjects(bundledDocuments, `/components/${componentsType}`);
+
+    if (Object.keys(mergedTypedComponents).length === 0) {
+      // Nothing was merged for this components type, go to the next component type
+      continue;
+    }
 
-  for (const resolvedRef of componentsMap.values()) {
-    insertRefByPointer(resolvedRef.pointer, resolvedRef.refNode, mergedComponents);
+    mergedComponents[componentsType] = mergedTypedComponents;
   }
 
   return mergedComponents;
 }
 
-function mergeRefsToMap(bundledRefs: ResolvedRef[], componentsMap: Map<string, ResolvedRef>): void {
-  for (const bundledRef of bundledRefs) {
-    const existingRef = componentsMap.get(bundledRef.pointer);
+function mergeObjects(
+  resolvedDocuments: ResolvedDocument[],
+  sourcePointer: string
+): Record<string, unknown> {
+  const merged: Record<string, unknown> = {};
+  const componentNameSourceLocationMap = new Map<string, string>();
+  const mergedEntityName = sourcePointer.split('/').at(-1);
 
-    if (!existingRef) {
-      componentsMap.set(bundledRef.pointer, bundledRef);
+  for (const resolvedDocument of resolvedDocuments) {
+    const object = extractObjectToMerge(resolvedDocument, sourcePointer);
+
+    if (!object) {
       continue;
     }
 
-    if (deepEqual(existingRef.refNode, bundledRef.refNode)) {
-      continue;
+    for (const name of Object.keys(object)) {
+      const componentToAdd = object[name];
+      const existingComponent = merged[name];
+
+      if (existingComponent && !deepEqual(componentToAdd, existingComponent)) {
```

Comment:
It turned out that sometimes it's convenient to have duplicated schema. For example `ExceptionListType` is defined in Detections API OpenAPI specs and in Exceptions API OpenAPI specs. Referencing a file locating quite far could lead to long relative paths. It can be mitigated by using packages for OpenAPI specs which was discussed in https://github.com/elastic/kibana/issues/184859.

I added logging a warning when duplicated components are encountered.